File: 2019-py/day14.py
#! /usr/bin/python3
# Copyright 2019 Google LLC
#
# Licensed under the Apache License, Version 2.0 (the "License");
# you may not use this file except in compliance with the License.
# You may obtain a copy of the License at
#
#      http://www.apache.org/licenses/LICENSE-2.0
#
# Unless required by applicable law or agreed to in writing, software
# distributed under the License is distributed on an "AS IS" BASIS,
# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
# See the License for the specific language governing permissions and
# limitations under the License.


import sys

# part 1

def pspec(spec):
    q, chem = spec.split(' ')
    return chem, int(q)

deps, quanta = {}, {}
with open('day14-input.txt' if len(sys.argv) < 2 else sys.argv[1]) as f:
    for line in f.readlines():
        inspecs, outspec = line.rstrip().split(' => ')
        outchem, outq = pspec(outspec)
        deps[outchem] = dict(pspec(inspec) for inspec in inspecs.split(', '))
        quanta[outchem] = outq

def tsort(deps):
    deps = dict((outchem, set(outdeps.keys())) for outchem, outdeps in deps.items())
    rdeps = {}
    for outchem, outdeps in deps.items():
        for dep in outdeps:
            rdeps.setdefault(dep, set()).add(outchem)
    l = []
    s = set(('FUEL',))
    while s:
        n = s.pop()
        l.append(n)
        while deps.get(n, []):
            m = deps[n].pop()
            rdeps[m].remove(n)
            if not rdeps[m]:
                s.add(m)
    return l

odeps = tsort(deps)

def ore(fuel):
    want = {'FUEL': fuel}
    for ch in odeps:
        if ch == 'ORE' or ch not in want: continue
        n = want.pop(ch)
        q = quanta[ch]
        k = (n + q - 1) // q
        for depch, depn in deps[ch].items():
            want[depch] = want.get(depch, 0) + k*depn
    return want['ORE']

print(ore(1))

# part 2

target = 1000000000000
start, end = 1, target+1

while end - start >= 2:
    mid = start + (end - start) // 2
    got = ore(mid)
    if got > target:
        end = mid
    else:
        start = mid

print(f'{start} -> {ore(start)}')
print(f'{end} -> {ore(end)}')

File: 2019/day14/day14.go
// Copyright 2019 Google LLC
//
// Licensed under the Apache License, Version 2.0 (the "License");
// you may not use this file except in compliance with the License.
// You may obtain a copy of the License at
//
//      http://www.apache.org/licenses/LICENSE-2.0
//
// Unless required by applicable law or agreed to in writing, software
// distributed under the License is distributed on an "AS IS" BASIS,
// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
// See the License for the specific language governing permissions and
// limitations under the License.

// Package day14 solves AoC 2019 day 14.
package day14

import (
	"fmt"
	"strconv"
	"strings"

	"github.com/fis/aoc/glue"
	"github.com/fis/aoc/util/fn"
	"github.com/fis/aoc/util/graph"
)

func init() {
	glue.RegisterSolver(2019, 14, glue.LineSolver(solve))
}

func solve(lines []string) ([]string, error) {
	reactions := parseReactions(lines)

	part1 := ore(1, reactions)
	part2 := maxFuel(1000000000000, reactions)

	return glue.Ints(part1, part2), nil
}

type pile struct {
	name string
	q    int
}

type reaction struct {
	out pile
	in  []pile
}

func ore(wantFuel int, reactions map[string]reaction) int {
	order := reactionOrder(reactions)
	want := map[string]int{"FUEL": wantFuel}
	return oreFor(want, order, reactions)
}

func maxFuel(ore int, reactions map[string]reaction) int {
	order := reactionOrder(reactions)
	start, end := 1, ore+1
	for end-start >= 2 {
		mid := start + (end-start)/2
		got := oreFor(map[string]int{"FUEL": mid}, order, reactions)
		if got > ore {
			end = mid
		} else {
			start = mid
		}
	}
	return start
}

func reactionOrder(reactions map[string]reaction) []string {
	gb := graph.NewBuilder()
	for out, r := range reactions {
		for _, in := range r.in {
			gb.AddEdgeL(out, in.name)
		}
	}
	g := gb.DenseDigraph()
	order := g.TopoSort(false)
	return fn.Map(order, g.Label)
}

func oreFor(want map[string]int, order []string, reactions map[string]reaction) int {
	for _, ch := range order {
		n, ok := want[ch]
		if !ok {
			continue // not needed
		}
		if ch == "ORE" {
			return n
		}
		delete(want, ch)
		r := reactions[ch]
		k := (n + r.out.q - 1) / r.out.q
		for _, in := range r.in {
			want[in.name] += k * in.q
		}
	}
	panic("no ore required")
}

func parseReactions(lines []string) map[string]reaction {
	reactions := make(map[string]reaction)
	for _, line := range lines {
		parts := strings.Split(line, " => ")
		if len(parts) != 2 {
			panic(fmt.Sprintf("invalid reaction: %s", line))
		}
		r := reaction{out: parsePile(parts[1])}
		for _, spec := range strings.Split(parts[0], ", ") {
			r.in = append(r.in, parsePile(spec))
		}
		reactions[r.out.name] = r
	}
	return reactions
}

func parsePile(spec string) pile {
	parts := strings.Split(spec, " ")
	if len(parts) != 2 {
		panic(fmt.Sprintf("invalid pile: %s", spec))
	}
	q, err := strconv.Atoi(parts[0])
	if err != nil {
		panic(fmt.Sprintf("invalid pile (not a number): %s", spec))
	}
	return pile{name: parts[1], q: q}
}
